controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentStatus.java

```diff
@@ -1181,28 +1202,69 @@ public Versions versions() {
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(type.zone(), versions, readyAt, change);
+            return Objects.hash(type.zone(), versions, readiness, change);
         }
 
         @Override
         public String toString() {
-            return change + " with versions " + versions + ", ready at " + readyAt;
+            return change + " with versions " + versions + ", " + readiness;
         }
 
     }
 
+    public enum DelayCause { none, unverified, notReady, blocked, running, coolingDown, invalidPackage, changeBlocked, paused }
+    public record Readiness(Instant at, DelayCause cause) implements Comparable<Readiness> {
+        public static final Readiness unverified = new Readiness(null, DelayCause.unverified);
+        public static final Readiness notReady = new Readiness(null, DelayCause.notReady);
+        public static final Readiness empty = new Readiness(Instant.EPOCH, DelayCause.none);
+        public Readiness(Instant at) { this(at, DelayCause.none); }
+        public Readiness blocked() { return new Readiness(at, DelayCause.blocked); }
+        public Readiness running() { return new Readiness(at, DelayCause.running); }
+        public boolean ok() { return at != null; }
+        public boolean okAt(Instant at) { return ok() && cause != DelayCause.running && cause != DelayCause.blocked && ! at.isBefore(this.at); }
+        @Override public int compareTo(Readiness o) {
+            return at == null ? o.at == null ?  0 : 1
+                              : o.at == null ? -1 : at.compareTo(o.at);
+        }
+        @Override public String toString() {
+            return ok() ? "ready at " + at + switch (cause) {
+                              case none -> "";
+                              case coolingDown -> ": cooling down after repeated failures";
+                              case blocked -> ": waiting for verification test to complete";
+                              case running -> ": waiting for current run to complete";
+                              case invalidPackage -> ": invalid application package, must resubmit";
+                              case changeBlocked -> ": deployment configuration blocks changes";
+                              case paused -> ": manually paused";
+                              default -> throw new IllegalStateException(cause + " should not have an instant at which it is ready");
+                          }
+                        : "not ready" + switch (cause) {
+                              case unverified -> ": waiting for verification test to complete";
+                              case notReady -> ": waiting for dependencies to complete";
+                              default -> throw new IllegalStateException(cause + " should have an instant at which it is ready");
+                          };
+        }
+    }
+
+    static <T extends Comparable<T>> T min(T a, T b) {
```

Comment:
FYI we have this in `vespajlib`: https://github.com/vespa-engine/vespa/blob/7e1b7baba3f2f723405985d636089650a521f5d7/vespajlib/src/main/java/com/yahoo/collections/Comparables.java